### ISSUE TITLE
Add a way to preserve a default export

### ISSUE THREAD

**Issue by Sytten** [rollup/rollup#4417]
### Feature Use Case

Hi!

Bit of context first, we have been trying to get Vue generic components to work using this blog post (https://logaretm.com/blog/generically-typed-vue-components/). The way vue works is by calling a function on the default export of the component to add a `render` function. 

In our code, we never use the default export so it seems like rollup removes it from the build despite us setting the `no-treeshake` option for the module. This is problematic because then the "side-effect" of Vue that adds the render function is never called.

### Feature Proposal

I am not really sure what it would look like to be honest as I only started working on rollup plugins today, but maybe it could be an option that transform/load/resolveId can set to prevent rollup from optimizing the exports.
Maybe it already exist, I just could not find it, I am sorry in advanced if that is the case.

Thanks you very much!

**Comment by lukastaegert** [comment]
I can only guess from your writing what your problem is.

> we never use the default export 

That sounds like you are searching for the export on the emitted chunk. If you want that, instead of playing with the `moduleSideEffect` flag, you should use `this.emitFile({id: <myModuleId>, type: 'chunk', fileName: <fileName>})`. This will expose a chunk that has all exports of the module.

**Comment by abeersaqib** [comment]
Hi @lukastaegert I am having more or less the same issue. I don't want my `default` export to be ripped off. 

My code 

```js
      class ABC extends EventTarget {
        constructor() {
          super();
        }
        setApi(o) {
          for (const k in o) {
            Object.defineProperty(this, k, {
              value: o[k],
            });
          }
          Object.freeze(this);
        }
      }
      const ABC = new ABC();
      
      export default ABC;
```


Expected Output is that I want to get is export default as it is. `export default ABC`;
while I am getting the `export {s as ABC}` which is causing all of my modules accessing the default to fail.

**Comment by RedMser** [comment]
I found a bit of a solution, though my setup is probably a bit different from what is described here.

I used `this.emitFile` to manually create chunks like described above. I also had to change `preserveEntrySignatures` to `allow-extension` in not only the `emitFile` call, but also on the root config level. This made the default exports stay for me.

**Comment by dnalborczyk** [comment]
@Sytten 

> My problem is that the default export is not present in the final build. The symbol seems to be stripped since it is not used.

it sounds like you expect the `default` export to exist in the bundle, while the component was rightfully deemed unused and stripped out? from what it sounds, that's contrary to how bundler work. if you want your component to be exported from the final bundle, you would need to re-export it from the entry file.

**Comment by dnalborczyk** [comment]
@Sytten @abeersaqib @RedMser it would be way more helpful if you would illustrate your issue with a reproduction, otherwise it's just a lot of guess work.

**Comment by Sytten** [comment]
I meant I wrote code above that you can use to reproduce. Yes I expect to be able to tell the bundler to keep some stuff since I do need it at runtime even if the bundler cant statically find that usage. I don't think this is an unreasonable requirement.

**Comment by dnalborczyk** [comment]
> I meant I wrote code above that you can use to reproduce. 

yes you did, and you left the remaining work to put that puzzle together for the maintainers.

copy/pasted your code. stripped out the type definitions.

https://rollupjs.org/repl/?version=2.72.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNsYXNzJTIwQ29tcG9uZW50JTIwJTdCJTVDbiUyMCUyMGRlZmluZSgpJTIwJTdCJTIwJTVDbiUyMCUyMCUyMCUyMHJldHVybiUyMCU3QiUyMCU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMkYlMkYlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbmNvbnN0JTIwbWFpbiUyMCUzRCUyMG5ldyUyMENvbXBvbmVudCgpLmRlZmluZSgpJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyME15Q29tcG9uZW50JTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlNUNuJTIwJTIwcmV0dXJuJTIwbWFpbiU1Q24lN0QlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjBtYWluJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

what are your expectations?

**Comment by dnalborczyk** [comment]
thank you for describing the issue in more depth.

>  I would need to check if the heuristics of rollup were improved recently and if that means it can now understand that it needs to keep the vue function, but last I checked it did not.

that would be great. there has been plenty of bug fixes and improvements, and I'm sure there's more to come.

> Basically I would like to be able to tell rollup to not "optimize" away this part of the code.

sounds reasonable.

**Comment by sapphi-red** [comment]
I think this is caused by the pure annotation injected by plugin-vue (https://github.com/vitejs/vite-plugin-vue/issues/11).
